crt0stack/src/lib.rs

```diff
@@ -0,0 +1,486 @@
+            Entry::EGid(v) => (EGid, v as _),
+            Entry::HWCap(v) => (HWCap, v as _),
+            Entry::ClockTick(v) => (ClockTick, v as _),
+            Entry::Secure(v) => (Secure, if v { 1 } else { 0 }),
+            Entry::HWCap2(v) => (HWCap2, v as _),
+        };
+        self.push_item(key)?;
+        self.push_item(value)?;
+        Ok(())
+    }
+
+    pub fn done(mut self) -> Result<Handle<'a>> {
+        self.push_item(Key::default())?;
+        self.push_item(Type::default())?;
+
+        let start_ptr = {
+            // at the end, copy the items of the data section from the bottom to the top of the stack
+            let (bottom, top) = self.stack.split_at_mut(self.items);
+
+            let (_prefix, src, _suffix) = unsafe { bottom.align_to_mut::<Type>() };
+
+            let (_prefix, dst, _suffix) = {
+                let end = self.data - self.items;
+                unsafe { top[0..end].align_to_mut::<Type>() }
+            };
+            let start = dst.len() - src.len();
+            let end = dst.len();
+            dst[start..end].copy_from_slice(src);
+            &dst[start] as *const usize as *const ()
+        };
+        Ok(Handle(self.stack, start_ptr))
+    }
+}
+
+impl<'a> Builder<'a, Env> {
+    // Saves the env string to the data section and creates
+    // a reference to it in the items section.
+    pub fn push(&mut self, env: &str) -> Result<()> {
+        self.save(0_u8)?; // c-str zero byte
+        let p = self.save_slice(env.as_bytes())?;
+        self.push_item(p)
+    }
+
+    pub fn next(mut self) -> Result<Builder<'a, Aux>> {
+        // last environ is NULL
+        self.push_item(0_usize)?;
+        Ok(Builder {
+            stack: self.stack,
+            data: self.data,
+            items: self.items,
+            state: PhantomData,
+        })
+    }
+}
+
+impl<'a> Builder<'a, Arg> {
+    pub fn new(stack: &'a mut [u8]) -> Self {
+        let len = stack.len();
+        Self {
+            stack,
+            data: len,
+            items: core::mem::size_of::<Type>(),
+            state: PhantomData,
+        }
+    }
+
+    // Saves the arg string to the data section and creates
+    // a reference to it in the items section.
+    // Uses `copy_within()` to shift the items section.
+    pub fn push(&mut self, arg: &str) -> Result<()> {
+        self.save(0_u8)?; // c-str zero byte
+        let p = self.save_slice(arg.as_bytes())?;
+        self.push_item(p)
+    }
+
+    pub fn next(mut self) -> Result<Builder<'a, Env>> {
+        // last argv is NULL
+        self.push_item(0_usize)?;
+
+        // Store argc at the beginning
+        let (prefix, dst, _suffix) = {
+            let start = 0;
+            let end = self.data;
+            unsafe { self.stack[start..end].align_to_mut::<Type>() }
+        };
+        if dst.is_empty() {
+            return Err(OutOfSpace);
+        }
+        dst[0] = ((self.items - prefix.len()) / core::mem::size_of::<Type>()) - 2; // argc
```

Comment:
This block of code is really similar to the other methods. Can we deduplicate somehow?

Reply:
not worth it